File: mnist/digit_classifier/random_model.py
from typing import Optional

import numpy as np

from .interface import DigitClassificationInterface


class RandomModel(DigitClassificationInterface):
    INPUT_SHAPE = (10, 10)

    def __init__(self, seed: Optional[int] = None) -> None:
        self._rng = np.random.default_rng(seed)

    def train(self, X: np.ndarray, y: np.ndarray) -> None:
        raise NotImplementedError("Training is out of scope for this exercise.")

    def predict(self, image: np.ndarray) -> int:
        if image.shape != self.INPUT_SHAPE:
            raise ValueError(f"RandomModel expects a {self.INPUT_SHAPE} array, got {image.shape}")

        return int(self._rng.integers(0, 10))

File: cpp/mnist/random_model.cpp
#include "random_model.hpp"

#include <stdexcept>

RandomModel::RandomModel(unsigned seed) : rng_(seed), dist_(0, 9) {}

void RandomModel::train(const std::vector<std::vector<double>>&, const std::vector<int>&) {
    throw std::logic_error("Training is out of scope for this exercise.");
}

int RandomModel::predict(const std::vector<double>& image) {
    if (image.size() != kInputLength) {
        throw std::invalid_argument("RandomModel::predict expects 100 values (a 10x10 crop)");
    }
    return dist_(rng_);
}
